src/sql/src/query_model/hir.rs

```diff
@@ -65,12 +65,34 @@ impl FromHir {
                 }))
             }
             HirRelationExpr::Map { input, scalars } => {
-                let box_id = self.generate_select(input);
-                // @todo self-referencing Maps
-                for scalar in scalars.iter() {
-                    let expr = self.generate_expr(scalar, box_id);
-                    let b = self.model.get_box(box_id);
-                    b.borrow_mut().columns.push(Column { expr, alias: None });
+                let mut box_id = self.generate_select(input);
+                let mut start_idx = 0;
+                let mut scalars = scalars.clone();
+                loop {
+                    let old_arity = self.model.get_box(box_id).borrow().columns.len();
+
+                    let end_idx = scalars
+                        .iter_mut()
+                        .position(|s| {
+                            let mut requires_nonexistent_column = false;
+                            s.visit_columns(0, &mut |depth, col| {
```

Comment:
I don't like cloning scalars here but the function `visit_columns` only accepts mutable references. The original `scalars` is an immutable reference since `generate_internal` accepts an immutable reference to the `HirRelationExpr`. Will accept any suggestions.

Reply:
Perhaps we could/should make the conversion to QGM _consume_ the HIR, so that the input relation can be mutable.